src/transformers/cache_utils.py

```diff
@@ -828,6 +829,7 @@ def __init__(self, config: PretrainedConfig, max_batch_size: int, max_cache_len:
             torch._dynamo.mark_static_address(new_layer_value_cache)
             self.key_cache.append(new_layer_key_cache)
             self.value_cache.append(new_layer_value_cache)
+            self.is_updated.append(False)
```

Comment:
Can we store this in the `EncoderDecoderCache` object instead? I see two potential benefits:
1. `StaticCache` would stay as simple as possible
2. The `is_updated` logic would work with all encoder cache types (e.g. dynamic cache and new cache types)

Reply:
I would be in favour of a design like this, but there's a small catch that means we'll have to make a compromise somewhere.

Note that we pass the `EncoderDecoderCache` wrapper to the decoder layer, but individual `Cache` objects to the attention layer (e.g. `self_attention_cache: Cache` to the self-attention layer, `cross_attention_cache: Cache` to the cross-attention layer)

Within the cross-attention layer, we need to know the value of `is_updated` (True or False). This means we either have to include `is_updated` in the `cross_attention_cache` class (as currently proposed), or add a new argument to the attention layer `is_updated`, which we set to the value of `EncoderDecoderCache(...).is_updated`:
```python
cross_attn_past_key_value = past_key_value.cross_attention_cache if past_key_value is not None else None
hidden_states, cross_attn_weights, cross_attn_present_key_value = self.encoder_attn(
    hidden_states=hidden_states,
    key_value_states=encoder_hidden_states,
    attention_mask=encoder_attention_mask,
    layer_head_mask=cross_attn_layer_head_mask,
    past_key_value=cross_attn_past_key_value,
    output_attentions=output_attentions,
+   is_updated=past_key_value.is_udpated if past_key_value is not None else None
)
```

Reply:
@sanchit-gandhi would it make sense to pass `EncoderDecoderCache` through the `past_key_value` argument and rely on `is_cross_attention` (`is_cross_attention = key_value_states is not None`) to select the correct cache inside the attention layer?

If the `EncoderDecoderCache` object could be passed, then it could hold `is_updated` without us needing an extra flag!

Reply:
Great idea @gante! Implemented in https://github.com/huggingface/transformers/pull/31166/commits/6d3997f08f639b6bf7d242c8a5dffbb74a105eb8. Let me know if you're happy with this design and more generally the state of the PR. If so, I can request a final review from Mr Zucker 🤗 Would be nice to have this one in for the next release!